libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,42 +112,121 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
         {
             if (ShouldSetToken(request))
             {
-                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", await GetTokenAsync());
+                string token = await this.GetTokenAsync().ConfigureAwait(false);
+                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
             }
-            await base.ProcessHttpRequestAsync(request, cancellationToken);
+            await base.ProcessHttpRequestAsync(request, cancellationToken).ConfigureAwait(false);
         }
 
         public async Task<string> GetTokenAsync(bool forceRefresh = false)
         {
-            if (forceRefresh == false)
+            Task<OAuthResponse> oAuthTokenTask = null;
+            OAuthResponse oAuthToken = null;
+
+            // get tokenTask from cache 
+            lock (tokenTaskCache)
```

Comment:
Any reason we don't change `tokeTaskCache` to a `ConcurrentDictionary` to avoid taking this explicit `lock`?

Reply:
because this lock is actually making sure that all of this code is only executed by one thread, it's not about the collections.

---
In reply to: [198878915](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198878915) [](ancestors = 198878915)